### DISCUSSION TITLE
@Dependency.AppStorage does not work when used with SwiftUI picker

### DISCUSSION THREAD

**Discussion by trispo** [tgrapperon/swift-dependencies-additions#37] [Q&A]
Hi,

I have a SwiftUI `Form` which shows a `Picker` which is driven by a `@Dependency.AppStorage` backed property in my model. The corresponding user default is set when changing the picker's selection, but the view is not updated accordingly. The picker is not dismissed and the checkmark does not change.

```swift
class ViewModel: ObservableObject {
    @Dependency.AppStorage("myDefault") public var myDefault = "one"
}

struct MyView: View {
    @StateObject var viewModel = ViewModel()

    var body: some View {
        NavigationStack {
            Form {
                Picker("Default value", selection: $viewModel.myDefault) {
                    Text("one").tag("one")
                    Text("two").tag("two")
                    Text("three").tag("three")
                }
            }
        }
    }
}
```

Any ideas?

Btw: I'm using watchOS.

**Comment by tgrapperon** [comment]
Hey @trispo! Thanks for the feedback.
Right now, `@Dependency.Appstorage` doesn't perform any magic in `ObservableObject` to automatically observe their value and call `objectWillChange`. As a result, changes are not automatically propagated at the SwiftUI level.

To make it work right now, you can assign this value to an honest `@Published` value that will call itself `objectWillChange`:
```swift
class ViewModel: ObservableObject {
  @Dependency.AppStorage("myDefault") public var myDefaultStorage = "one"
  @Published var myDefault: String = ""
  init() {
    $myDefaultStorage.assign(to: &$myDefault)
  }
}
```
This is how it is performed in the case studies app.

Alternatively, you can also iterate over this dependency's `projectedValue` (using `$myDefault`) and call `objectWillChange` yourself:
```swift
class ViewModel: ObservableObject {
  @Dependency.AppStorage("myDefault") public var myDefault = "one"
  var cancellables: Set<AnyCancellableTask> = []
  init() {
    Task { [weak self] in
      guard let self else { return }
      for await _ in self.$myDefault {
        self.objectWillChange.send()
      }
    }.store(in: &cancellables)
  }
}
```
The reason it works like this right now is that this dependency was designed with TCA in mind, where observation would be explicit. That being said, it is maybe possible to tweak `AppStorage`'s implementation so it calls `objectWillChange` automatically when in an `ObservableObject`. I'll see if I can make it work and I'll keep you posted!

**Reply by trispo** [reply]
Great, thanks for your reply. Will use it as suggested for now.
